Source/Scene/ModelExperimental/InstancingPipelineStage.js

```diff
@@ -0,0 +1,272 @@
+import Cartesian3 from "../../Core/Cartesian3.js";
+import ComponentDatatype from "../../Core/ComponentDatatype.js";
+import defined from "../../Core/defined.js";
+import Matrix4 from "../../Core/Matrix4.js";
+import AttributeType from "../AttributeType.js";
+import Quaternion from "../../Core/Quaternion.js";
+import Buffer from "../../Renderer/Buffer.js";
+import BufferUsage from "../../Renderer/BufferUsage.js";
+import InstanceAttributeSemantic from "../InstanceAttributeSemantic.js";
+import ModelExperimentalUtility from "./ModelExperimentalUtility.js";
+import InstancingStageVS from "../../Shaders/ModelExperimental/InstancingStageVS.js";
+
+/**
+ * The instancing pipeline stage is responsible for handling GPU mesh instancing at the node
+ * level.
+ *
+ * @namespace
+ * @private
+ */
+var InstancingPipelineStage = {};
+
+/**
+ * Process a node. This modifies the following parts of the render resources:
+ * <ul>
+ *  <li>adds attribute declarations for the instancing vertex attributes in the vertex shader</li>
+ *  <li>adds an instancing translation min and max to compute an accurate bounding volume</li>
+ * </ul>
+ * @param {RenderResources.NodeRenderResources} renderResources The render resources for this node.
+ * @param {ModelComponents.Node} node The node.
+ * @param {FrameState} frameState The frame state.
+ */
+InstancingPipelineStage.process = function (renderResources, node, frameState) {
+  var instances = node.instances;
+  var instancingVertexAttributes = [];
+
+  var shaderBuilder = renderResources.shaderBuilder;
+  shaderBuilder.addDefine("HAS_INSTANCING");
+  shaderBuilder.addVertexLines([InstancingStageVS]);
+
+  var translationAttribute = ModelExperimentalUtility.getAttributeBySemantic(
+    instances,
+    InstanceAttributeSemantic.TRANSLATION
+  );
+
+  var translationMax;
+  var translationMin;
+  if (defined(translationAttribute)) {
+    translationMax = translationAttribute.max;
+    translationMin = translationAttribute.min;
+  }
+
+  var rotationAttribute = ModelExperimentalUtility.getAttributeBySemantic(
+    instances,
+    InstanceAttributeSemantic.ROTATION
+  );
+  if (
+    defined(rotationAttribute) ||
+    (!defined(translationMax) && !defined(translationMin))
+  ) {
+    var transformsTypedArray = getInstanceTransformsTypedArray(
+      instances,
+      renderResources
+    );
+    var transformsVertexBuffer = Buffer.createVertexBuffer({
```

Comment:
Is this a resource that needs to be `destroy()`ed at some point in time? This gets back to our outstanding question of how to manage resources created by pipeline stages.

Reply:
I implemented a really simple system where the `ModelExperimental` keeps an array of `_resources` that need to be destroyed. This can be easily accessed since the `RenderResources` contain a reference to the `ModelExperimental`

Reply:
See https://github.com/CesiumGS/cesium/pull/9713/commits/95720cd204982076528f74bc40371381b95924d5